signwriting_evaluation/metrics/symbol_distance.py

```diff
@@ -0,0 +1,93 @@
+
+import numpy as np
+from scipy.optimize import linear_sum_assignment
+from scipy.spatial import distance as dis
+from signwriting.types import Sign, SignSymbol
+from signwriting.formats.fsw_to_sign import fsw_to_sign
+from signwriting_evaluation.metrics.base import SignWritingMetric
+
+
+class SignWritingSimilarityMetric(SignWritingMetric):
+    def __init__(self):
+        super().__init__("SymbolsDistances")
+        self.symbol_classes = {
+            'hand_shapes': range(0x100, 0x205),
+            'contact_symbols': range(0x205, 0x221),
+            'movement_paths': range(0x221, 0x2FF),
+            'head_movement': range(0x2FF, 0x30A),
+            'facial_expressions': range(0x30A, 0x36A),
+            'etc': range(0x36A, 0x38C)
+        }
+        self.weight = {
+            "shape": 24,  # same weight as switching parallelization
+            "facing": 8,  # one column (facing) equals 8 out of 16 lines of the table (like switching hands)
+            "angle": 1,  # lowest importance out of the criteria
+            "parallel": 24,  # parallelization is 3 columns compare to 1 for the facing direction
+            "positional": 1,  # may be big numerical differance, balancing with 1
+            "normalized_factor": 1 / 5,  # fitting shape of function
+            "class_penalty": 84,  # big penalty for each class type passed
+        }
+        self.max_distance = self.calculate_distance({"symbol": "S10000", "position": (250, 250)},
+                                                    {"symbol": "S38b07", "position": (750, 750)})
+
+    def euc_distance(self, first: Tuple[int, int], second: Tuple[int, int]) -> float:
+        return sqrt(pow(first[0] - second[0], 2) + pow(first[1] - second[1], 2))
+
+    def get_attributes(self, symbol: SignSymbol) -> Tuple[int, int, int, bool]:
+        shape = int(symbol['symbol'][1:4], 16)
+        facing = int(symbol['symbol'][4], 16)
+        angle = int(symbol['symbol'][5], 16)
+        parallel = facing > 2  # left or right of sign table (0 vertical, 1 horizontal)
+        return shape, facing, angle, parallel
+
+    def calculate_distance(self, hyp: SignSymbol, ref: SignSymbol) -> float:
+        shape1, angle1, facing1, parallel1 = self.get_attributes(hyp)
+        shape2, angle2, facing2, parallel2 = self.get_attributes(ref)
+        distance = (self.weight["shape"] * abs(shape1 - shape2) +
+                    self.weight["facing"] * abs(facing1 - facing2) +
+                    self.weight["angle"] * abs(angle1 - angle2) +
+                    self.weight["parallel"] * (parallel1 != parallel2) +
+                    self.weight["positional"] * dis.euclidean(hyp["position"], ref["position"]))
+        hyp_class = next((i for i, r in enumerate(self.symbol_classes.values()) if shape1 in r), None)
+        ref_class = next((i for i, r in enumerate(self.symbol_classes.values()) if shape2 in r), None)
+        distance = distance + abs(ref_class - hyp_class) * self.weight["class_penalty"]
```

Comment:
i feel like this might be too lenient - in my opinion, same class = 1, different class = 0